### DISCUSSION TITLE
Reliability of file libopensc.exports

### DISCUSSION THREAD

**Discussion by carblue** [OpenSC/OpenSC#3255] [General]
Hello,
I'm author of an external driver that links dynamically to libopensc.so/.dll etc. Thus it's crucial for me to know, what exactly gets exported from  libopensc.so, and in the past I thought that file libopensc.exports serves that purpose.
I was wrong: E.g. since version 0.22.0 function sc_compare_oid is not contained anymore in file libopensc.exports, but in fact,
`$ nm -D /usr/lib/x86_64-linux-gnu/libopensc.so.12.0.0 | grep sc_compare_oid`
shows, that the function still is in libopensc.so.12.0.0 (checked with opensc version 0.26.0-rc1).

Isn't file libopensc.exports  superfluous then?

**Comment by dengert** [comment]
It is in https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/libopensc.exports#L67
and git blame saws for 17 years:
https://github.com/OpenSC/OpenSC/blame/master/src/libopensc/libopensc.exports#L67:~:text=sc_check_sw-,17%20years%20ago,sc_compare_oid,-sc_compare_path

How are you using `libopensc.exports`? Its not installed so you have to get it from the source directory. Are you building OpenSC in separate directory from the source?

So you maybe seeing a distro change in recent years
and you may have to install the `-dev package (if they have one for OpenSC) to get an unstriped version. 
The OpenSC really only exports PKCS11 and utilities, libopensc is meant to be internal. 

On Ubuntu 22.04 when I build and install OpenSC for testing with debugging its there:
```
$ nm -D /opt/ossl-3.3.1/lib/libopensc.so.12.0.0 | grep sc_compare_oid
00000000000217d4 T sc_compare_oid
$ file /opt/ossl-3.3.1/lib/libopensc.so.12.0.0
/opt/ossl-3.3.1/lib/libopensc.so.12.0.0: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=d236086d73025fad0afde3afae7c84259b26fe3d, with debug_info, not stripped
```

But the system says:
```
$ nm /usr/lib/x86_64-linux-gnu/libopensc.so.11.0.0
nm: /usr/lib/x86_64-linux-gnu/libopensc.so.11.0.0: no symbols
$ file /usr/lib/x86_64-linux-gnu/libopensc.so.11.0.0
/usr/lib/x86_64-linux-gnu/libopensc.so.11.0.0: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=10b7eed1a1edfc4821d7a07a7f80a2143be405e6, stripped
```

**Comment by carblue** [comment]
@dengert 
Sorry, I wasn't careful enough!  It is included as You state.

I stumbled upon the fact that function sc_compare_oid was included twice in version 0.21.0 libopensc.exports, and (in the first place) only saw the (correct) deletion of the duplicate entry beginning with version 0.22.0. (An alphabetical sorting within the libopensc.exports file would have helped prevent my error from occurring in the first place)

It might have been obscure what I'm doing here with libopensc.exports, old OpenSC versions, libopensc.so  etc.,
From my point of view it's natural: I decided to learn the Rust programming language along a hands-on project: An OpenSC driver for ACOS5 smart cards, https://github.com/carblue/acos5
I'm really glad, that OpenSC allows external drivers, and my one is able to support multiple OpenSC versions since 0.20.0 (as Linux distros often lack behind the latest release). For this to work, while building the driver, it detects the OpenSC version installed, and adapts it's Rust "header"/binding files like e.g. opensc.rs to exactly that version by conditional compilation.
Thus I deal a lot with OpenSC C header files, libopensc.so etc. what You think is meant for internal use only.
OpenSC: I build that for my Kubuntu 24.04 from source tarballs (or git clone) acc.  https://github.com/OpenSC/OpenSC/wiki/Compiling-and-Installing-on-Unix-flavors, only deviating with  --libdir=/usr/lib/x86_64-linux-gnu, and finally sudo checkinstall

I think, I should ignore file libopensc.exports and rely only on what
`$ nm -D /usr/lib/x86_64-linux-gnu/libopensc.so`
tells me

**Comment by frankmorgner** [comment]
Since many releases now, we don't make any guarantees regarding linkage against libopensc. This library should only be used by OpenSC tools and drivers. If you want to get stable support for your card you will have to get the driver integrated upstream.

**Comment by carblue** [comment]
I understand Your point of view and don't expect any guarantees regarding linkage against libopensc.
My point of view is different, obviously not understood: "My" ACOS5 driver **only** exists, because there is a way to write it in Rust, not C. It's my time, that I spend permanently since many releases now to make the Rust FFI working for linkage against libopensc.

For me, this discussion was closed as resolved. Did I miss to state so? Sorry.

**Reply by frankmorgner** [reply]
I see. Thanks for the update. Maybe you want to go the extra mile and implement the PKCS#11 interface (or even Windows Minidriver and Apple's CTK). Those are stable interfaces and if you have a specific card you can take many shortcuts which in the end don't add much overhead compared to maintaining a separated OpenSC card driver.
